source/view.c

```diff
@@ -540,25 +553,95 @@ static void rofi_view_call_thread ( gpointer data, gpointer user_data )
     g_mutex_unlock ( t->mutex );
 }
 
+enum CharClass { LOWER, UPPER, DIGIT, NON_WORD };
+
+static enum CharClass char_class(gunichar c)
+{
+  if (g_unichar_islower(c))
+    return LOWER;
+  if (g_unichar_isupper(c))
+    return UPPER;
+  if (g_unichar_isdigit(c))
+    return DIGIT;
+  return NON_WORD;
+}
+
+static int score_for(enum CharClass prev, enum CharClass curr)
+{
+  if (prev == NON_WORD && curr != NON_WORD)
+    return WORD_START_SCORE;
+  if (prev == LOWER && curr == UPPER ||
+      prev != DIGIT && curr == DIGIT)
+    return CAMEL_SCORE;
+  if (curr == NON_WORD)
+    return NON_WORD_SCORE;
+  return 0;
+}
+
 static void filter_elements ( thread_state *t, G_GNUC_UNUSED gpointer user_data )
 {
-    // input changed
+    char *pattern = mode_preprocess_input(t->state->sw, t->state->text->text);
+    glong plen = g_utf8_strlen(pattern, -1);
     for ( unsigned int i = t->start; i < t->stop; i++ ) {
         int match = mode_token_match ( t->state->sw, t->state->tokens, i );
         // If each token was matched, add it to list.
         if ( match ) {
             t->state->line_map[t->start + t->count] = i;
-            if ( config.levenshtein_sort ) {
+            if (config.matching_method == MM_FUZZY) {
+                char *str = mode_get_completion(t->state->sw, i);
+                glong slen = g_utf8_strlen(str, -1), pi, si;
+                if (slen > 256)
+                    t->state->distance[i] = MIN_SCORE;
+                else {
+                    gboolean pfirst = TRUE;
+                    int *score = g_malloc_n(slen, sizeof(int)),
+                        *dp = g_malloc_n(slen, sizeof(int)),
+                        *dps = g_malloc_n(slen, sizeof(int));
+                    gchar *pit = pattern, *sit;
+                    enum CharClass prev = NON_WORD, cur;
+                    for (si = 0, sit = str; si < slen; si++, sit = g_utf8_next_char(sit)) {
+                        cur = char_class(g_utf8_get_char(sit));
+                        score[si] = score_for(prev, cur);
+                        prev = cur;
+                    }
+                    for (pi = 0; pi < plen; pi++, pit = g_utf8_next_char(pit)) {
+                        int prefix = MIN_SCORE, ul, ul2, uls, uls2;
+                        if (g_utf8_get_char(pit) == ' ')
```

Comment:
Rofi coding style is to always add {} .
(there is a make indent target  you can run if you have uncrustify installed).